tests/link/test_numba_performance.py

```diff
@@ -0,0 +1,80 @@
+import time
+
+import numpy as np
+import pytest
+
+import aesara.tensor as aet
+from aesara import config
+from aesara.compile.function import function
+from aesara.compile.mode import Mode
+from aesara.compile.sharedvalue import SharedVariable
+from aesara.graph.basic import Constant
+from aesara.graph.fg import FunctionGraph
+from aesara.graph.optdb import OptimizationQuery
+from aesara.link.numba.linker import NumbaLinker
+
+
+opts = OptimizationQuery(include=[None], exclude=["cxx_only", "BlasOpt"])
+numba_mode = Mode(NumbaLinker(), opts)
+py_mode = Mode("py", opts)
+
+
+def set_test_value(x, v):
+    x.tag.test_value = v
+    return x
+
+
+def run_numpy_benchmark(fgraph, numpy_fn, inputs, numpy_inputs, performace_factor=1):
+    fn_inputs = [i for i in fgraph.inputs if not isinstance(i, SharedVariable)]
+    aesara_numba_fn = function(
+        fn_inputs,
+        fgraph.outputs,
+        mode=numba_mode,
+        accept_inplace=True,
+    )
+    numba_res = aesara_numba_fn(*inputs)
+    numpy_res = numpy_fn(*numpy_inputs)
+
+    assert np.all(numba_res == numpy_res)
+
+    t0 = time.time()
+    aesara_numba_fn(*inputs)
+    t1 = time.time()
+    numpy_fn(*numpy_inputs)
+    t2 = time.time()
+
+    assert t1 - t0 <= performace_factor * (t2 - t1)
+
+
+@pytest.mark.parametrize(
+    "careduce_fn, numpy_fn, aesara_inputs, numpy_inputs",
+    [
+        (
+            aet.sum,
+            np.sum,
+            [
+                set_test_value(
+                    aet.matrix(),
+                    np.arange(3000 * 2000, dtype=config.floatX).reshape((3000, 2000)),
+                ),
+                None,
+            ],
+            [np.arange(3000 * 2000, dtype=config.floatX).reshape((3000, 2000)), None],
+        )
+    ],
+)
+def test_careduce_performance(careduce_fn, numpy_fn, aesara_inputs, numpy_inputs):
+    g = careduce_fn(*aesara_inputs)
+    g_fg = FunctionGraph(outputs=[g])
+
+    run_numpy_benchmark(
+        g_fg,
+        numpy_fn,
+        [
+            i.tag.test_value
+            for i in g_fg.inputs
+            if not isinstance(i, (SharedVariable, Constant))
+        ],
+        numpy_inputs,
+        performace_factor=4,
```

Comment:
Looks like the performance isn't that good with this implementation either.

Reply:
We might need to pre-compile those `CAReduce` loops using [`use_optimized_cheap_pass`](https://gist.github.com/brandonwillard/0302f061fc9f89a65b5517170347a6f0#file-numba-careduce-mwe-py-L96).

Reply:
I just pushed changes adding that. But it doesn't look like it's making any difference in performance, can you confirm if it's correctly implemented ? (with respect to the `exec` interface)

Reply:
The context manager needs to be used when Numba JIT compiles a function, but it looks like it's been added to the Python source string compilation utility.